node/core/pvf/src/execute/queue.rs

```diff
@@ -72,12 +72,48 @@ struct ExecuteJob {
 	waiting_since: Instant,
 }
 
+enum WorkerState {
+	Idle(IdleWorker),
+	Assigned(ExecuteJob),
+	Busy,
+}
+
 struct WorkerData {
-	idle: Option<IdleWorker>,
+	state: WorkerState,
 	handle: WorkerHandle,
 	executor_params_hash: ExecutorParamsHash,
 }
 
+impl WorkerData {
+	/// Assigns a job to the worker. Returns the idle worker token. An attempt to assign a job to
+	/// a worker already busy with another job means there's a critical bug in code and shall
+	/// result in a panic.
+	fn assign(&mut self, job: ExecuteJob) -> IdleWorker {
+		debug_assert_eq!(self.executor_params_hash, job.executor_params.hash());
+		match mem::replace(&mut self.state, WorkerState::Assigned(job)) {
+			WorkerState::Idle(idle) => idle,
+			_ => unreachable!("Tried to assign a job to already busy worker"),
+		}
+	}
+
+	/// Runs the assigned job. Returns job data to pass to the worker.
+	fn run(&mut self) -> ExecuteJob {
```

Comment:
We already call any spawned worker a "running" worker, so overloading the term `run` here is a bit confusing. But this may not be needed in addition to `assign`.